vllm/model_executor/models/whisper.py

```diff
@@ -0,0 +1,689 @@
+    audio, orig_sr = multi_modal_data["audio"]
+    processor = get_whisper_processor(ctx.model_config.model)
+    target_sr = processor.feature_extractor.sampling_rate
+    audio = librosa.resample(audio, orig_sr=orig_sr, target_sr=target_sr)
+    if audio.size > 30 * target_sr:
+        # Truncate audio to 30 seconds
+        audio = audio[:30 * target_sr]
+    multi_modal_data["audio"] = (audio, target_sr)
+    # Calculate number of tokens after convolutions
+    num_tokens = (audio.size // 80 - 1) // 2 + 1
+    num_tokens = (num_tokens - 2) // 2 + 1
+    # Pre-allocate placeholder tokens in encoder sequence
+    inputs["encoder"]["prompt_token_ids"] = [0] * num_tokens
+    return inputs
+
+
+def input_mapper_for_whisper(
+    ctx: InputContext,
+    multi_modal_data: Union[np.ndarray, List[np.ndarray]],
+) -> MultiModalKwargs:
+    if not isinstance(multi_modal_data, list):
+        multi_modal_data = [multi_modal_data]
+
+    assert len(multi_modal_data) == 1
+
+    if len(multi_modal_data) == 0:
+        return MultiModalKwargs()
+
+    processor = get_whisper_processor(ctx.model_config.model)
+    sampling_rate = processor.feature_extractor.sampling_rate
+
+    audios = [audio for audio, _ in multi_modal_data]
+
+    kwargs = processor(audios, sampling_rate=sampling_rate,
+                       padding=False, return_tensors="pt")
+    kwargs["input_features"] = kwargs["input_features"].squeeze(0)
+    kwargs["input_features"] = kwargs["input_features"].to(torch.float16)
+
+    return MultiModalKwargs(kwargs)
+
+
+@INPUT_REGISTRY.register_dummy_encoder_data(dummy_encoder_data_for_whisper)
+@INPUT_REGISTRY.register_input_processor(input_processor_for_whisper)
+@MULTIMODAL_REGISTRY.register_input_mapper("audio", input_mapper_for_whisper)
+@MULTIMODAL_REGISTRY.register_max_multimodal_tokens(
+    "audio", get_max_whisper_audio_tokens)
+class WhisperForConditionalGeneration(nn.Module, SupportsMultiModal):
+
+    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
+        super().__init__()
+        config = vllm_config.model_config.hf_config
+        quant_config = vllm_config.quant_config
+        multimodal_config = vllm_config.model_config.multimodal_config
+        self.config = config
+        self.dtype = vllm_config.model_config.dtype
+
+        self.model = WhisperModel(vllm_config=vllm_config, prefix=prefix)
+        self.unpadded_vocab_size = config.vocab_size
+        self.proj_out = ParallelLMHead(config.vocab_size,
+                                       config.d_model,
+                                       quant_config=quant_config)
+        self.proj_out = self.proj_out.tie_weights(
+            self.model.decoder.embed_tokens)
+        logit_scale = getattr(config, "logit_scale", 1.0)
+        self.logits_processor = LogitsProcessor(self.unpadded_vocab_size,
+                                                config.vocab_size, logit_scale)
+        self.sampler = Sampler()
+
+    def forward(
```

Comment:
To support V1, the model needs to implement `get_input_embeddings` and `get_multimodal_embeddings`.

Reply:
Done, but there is some awkwardness
- `get_multimodal_embeddings` for Whisper also requires kv-cache and attn metadata which is not in the base method's interface
- It looks like `get_input_embeddings` expects to embed multimodal tokens into text tokens. again, this is not what Whisper does. Currently I just have this method return the decoder token embeddings.

I couldn't find anywhere in vLLM that uses these two methods so it's hard for me to understand their intended usage. Happy to change the implementation to whatever works better.

Reply:
Sorry, I forgot that V1 doesn't support encoder decoder yet. Let's just add a TODO and come back to this later.

Reply:
added todos